### PR TITLE
Add docker command to point a tag to manifest

### PR DESCRIPTION
**Pull request by daviddavis**
Adding a command that will update a tag to point a manifest given a digest. If
the tag does not exist, this command will create it. The options are repo id,
tag name, and manifest digest.

closes #2189
https://pulp.plan.io/issues/2189

# TODO
- [x] Add tests
- [x] Add validation for the digest field (make sure the manifest exists with the supplied digest)
- [x] Add release notes
- [x] Update docs

### CODE REVIEW

**Comment by dkliban** [comment]
ok test

File: extensions_admin/pulp_docker/extensions/admin/upload.py [hunk 1]
```diff
@@ -64,3 +72,69 @@ def generate_override_config(self, **kwargs):
...
+        :type  kwargs: dict
+
+        :return: unit key that should be uploaded for the file
+        :rtype:  dict
+        """
+
+        unit_key = {}
+        metadata = {}
+
+        return unit_key, metadata
```
Comment (dkliban) [inline extensions_admin/pulp_docker/extensions/admin/upload.py]: This method should return a unit key consisting of tag name and repo_id. It's defined here https://github.com/pulp/pulp_docker/blob/master/plugins/pulp_docker/plugins/models.py#L280

Comment (dkliban) [inline extensions_admin/pulp_docker/extensions/admin/upload.py]: This should only be the unit_key in following format {'name: 'blah', 'repo_id': 'blah'}

**Comment by ipanova** [comment]
@daviddavis  talking about validation, i think besides checking manifest digest, we should also check repo existence

**Comment by daviddavis** [comment]
@ipanova it looks like the code already validates that the repo exists that corresponds to `--repo-id`. Do you mean validate that the manifest with the supplied digest is in the repo?

**Comment by ipanova** [comment]
@daviddavis yes
```$ pulp-admin docker repo remove manifest --repo-id synctest --str-eq='digest=sha256:2824fe048727a69da66cf1be00cebd3bb9cfe1f238473693aa9358b411208527'

This command may be exited via ctrl+c without affecting the request.

[\]
Running...

Units Removed:
  sha256:2824fe048727a69da66cf1be00cebd3bb9cfe1f238473693aa9358b411208527
$ pulp-admin docker repo list 
Warning: path should have mode 0700 because it may contain sensitive information: /home/ipanova/.pulp/

+----------------------------------------------------------------------+
                          Docker Repositories
+----------------------------------------------------------------------+

Id:                  synctest
Display Name:        None
Description:         None
Content Unit Counts: 
  Docker Blob:     18
  Docker Manifest: 42
  Docker Tag:      43

Id:                  test
Display Name:        None
Description:         None
Content Unit Counts: 
  Docker Blob:     2
  Docker Manifest: 1
  Docker Tag:      1


$ pulp-admin docker repo tag --repo-id synctest --tag-name=glibc-test --manifest-digest='sha256:2824fe048727a69da66cf1be00cebd3bb9cfe1f238473693aa9358b411208527'
Warning: path should have mode 0700 because it may contain sensitive information: /home/ipanova/.pulp/

+----------------------------------------------------------------------+
                              Unit Upload
+----------------------------------------------------------------------+

Extracting necessary metadata for each request...
... completed

Creating upload requests on the server...
[==================================================] 100%
Initializing upload
... completed

Starting upload of selected units. If this process is stopped through ctrl+c,
the uploads will be paused and may be resumed later using the resume command or
canceled entirely using the cancel command.

Importing into the repository...
This command may be exited via ctrl+c without affecting the request.


[\]
Running...

Task Succeeded


Deleting the upload request...
... completed

[ipanova@ina pulp_docker]$ pulp-admin docker repo list 
Warning: path should have mode 0700 because it may contain sensitive information: /home/ipanova/.pulp/

+----------------------------------------------------------------------+
                          Docker Repositories
+----------------------------------------------------------------------+

Id:                  synctest
Display Name:        None
Description:         None
Content Unit Counts: 
  Docker Blob:     18
  Docker Manifest: 42
  Docker Tag:      44

Id:                  test
Display Name:        None
Description:         None
Content Unit Counts: 
  Docker Blob:     2
  Docker Manifest: 1
  Docker Tag:      1

```

Notice the count of the tag 43--->44, that means we basically added a tag of a manifest digest that does not exist in this repo.

**Comment by daviddavis** [comment]
@ipanova will do.

**Comment by ipanova** [comment]
@daviddavis also, could you please open a pulp-smash issue with some test scenarios?

**Comment by ipanova** [comment]
@daviddavis seems like the docker error is not propagated correctly
```
$ pulp-admin -vv docker repo tag --repo-id test2 --tag-name=glibc-test --manifest-digest='sha256:1111111111'
Warning: path should have mode 0700 because it may contain sensitive information: /home/ipanova/.pulp/

+----------------------------------------------------------------------+
                              Unit Upload
+----------------------------------------------------------------------+

2016-11-16 13:09:49,659 - DEBUG - sending GET request to /pulp/api/v2/repositories/test2/
2016-11-16 13:09:55,217 - INFO - GET request to /pulp/api/v2/repositories/test2/ with parameters None
2016-11-16 13:09:55,217 - INFO - Response status : 200 

2016-11-16 13:09:55,217 - INFO - Response body :
 {
  "scratchpad": {}, 
  "display_name": null, 
  "description": null, 
  "last_unit_added": "2016-11-16T11:09:30Z", 
  "notes": {
    "_repo-type": "docker-repo"
  }, 
  "last_unit_removed": null, 
  "content_unit_counts": {
    "docker_blob": 2, 
    "docker_manifest": 2, 
    "docker_tag": 2
  }, 
  "_ns": "repos", 
  "_id": {
    "$oid": "582c3e5945ef48698a672283"
  }, 
  "id": "test2", 
  "_href": "/pulp/api/v2/repositories/test2/"
}

Extracting necessary metadata for each request...
... completed

Creating upload requests on the server...
[==================================================] 100%
Initializing upload
2016-11-16 13:09:55,218 - DEBUG - sending POST request to /pulp/api/v2/content/uploads/
2016-11-16 13:09:55,306 - INFO - POST request to /pulp/api/v2/content/uploads/ with parameters None
2016-11-16 13:09:55,306 - INFO - Response status : 201 

2016-11-16 13:09:55,306 - INFO - Response body :
 {
  "upload_id": "68ca05ad-468e-445d-bb83-fd1875058d0a", 
  "_href": "/pulp/api/v2/content/uploads/68ca05ad-468e-445d-bb83-fd1875058d0a/"
}

... completed

Starting upload of selected units. If this process is stopped through ctrl+c,
the uploads will be paused and may be resumed later using the resume command or
canceled entirely using the cancel command.

Importing into the repository...
2016-11-16 13:09:55,307 - DEBUG - sending POST request to /pulp/api/v2/repositories/test2/actions/import_upload/
2016-11-16 13:09:55,470 - INFO - POST request to /pulp/api/v2/repositories/test2/actions/import_upload/ with parameters {"override_config": {}, "unit_type_id": "docker_tag", "upload_id": "68ca05ad-468e-445d-bb83-fd1875058d0a", "unit_key": {"repo_id": "test2", "name": "glibc-test"}, "unit_metadata": {"name": "glibc-test", "digest": "sha256:1111111111"}}
2016-11-16 13:09:55,470 - INFO - Response status : 202 

2016-11-16 13:09:55,470 - INFO - Response body :
 {
  "spawned_tasks": [
    {
      "_href": "/pulp/api/v2/tasks/664c3889-cf77-4235-bd44-2161e500ffbb/", 
      "task_id": "664c3889-cf77-4235-bd44-2161e500ffbb"
    }
  ], 
  "result": null, 
  "error": null
}

This command may be exited via ctrl+c without affecting the request.


[-]
Running...
2016-11-16 13:09:56,472 - DEBUG - sending GET request to /pulp/api/v2/tasks/664c3889-cf77-4235-bd44-2161e500ffbb/
2016-11-16 13:09:56,557 - INFO - GET request to /pulp/api/v2/tasks/664c3889-cf77-4235-bd44-2161e500ffbb/ with parameters None
2016-11-16 13:09:56,557 - INFO - Response status : 200 

2016-11-16 13:09:56,557 - INFO - Response body :
 {
  "exception": null, 
  "task_type": "pulp.server.managers.content.upload.import_uploaded_unit", 
  "_href": "/pulp/api/v2/tasks/664c3889-cf77-4235-bd44-2161e500ffbb/", 
  "task_id": "664c3889-cf77-4235-bd44-2161e500ffbb", 
  "tags": [
    "pulp:repository:test2", 
    "pulp:action:import_upload"
  ], 
  "finish_time": "2016-11-16T12:09:55Z", 
  "_ns": "task_status", 
  "start_time": "2016-11-16T12:09:55Z", 
  "traceback": "Traceback (most recent call last):\n  File \"/usr/lib/python2.7/site-packages/celery/app/trace.py\", line 240, in trace_task\n    R = retval = fun(*args, **kwargs)\n  File \"/home/ipanova/pulp_development/pulp/server/pulp/server/async/tasks.py\", line 488, in __call__\n    return super(Task, self).__call__(*args, **kwargs)\n  File \"/home/ipanova/pulp_development/pulp/server/pulp/server/async/tasks.py\", line 103, in __call__\n    return super(PulpTask, self).__call__(*args, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/celery/app/trace.py\", line 437, in __protected_call__\n    return self.run(*args, **kwargs)\n  File \"/home/ipanova/pulp_development/pulp/server/pulp/server/managers/content/upload.py\", line 223, in import_uploaded_unit\n    unit_type=unit_type_id, summary=result['summary'], details=result['details']\nPulpCodedException: The importer docker_importer indicated a failed response when uploading docker_tag unit to repository test2.\n", 
  "spawned_tasks": [], 
  "progress_report": {}, 
  "queue": "reserved_resource_worker-0@ina.usersys.redhat.com.dq", 
  "state": "error", 
  "worker_name": "reserved_resource_worker-0@ina.usersys.redhat.com", 
  "result": null, 
  "error": {
    "code": "PLP0047", 
    "data": {
      "unit_type": "docker_tag", 
      "importer_id": "docker_importer", 
      "repo_id": "test2", 
      "details": {}, 
      "summary": ""
    }, 
    "description": "The importer docker_importer indicated a failed response when uploading docker_tag unit to repository test2.", 
    "sub_errors": []
  }, 
  "_id": {
    "$oid": "582c4c93e20871aff6e898e7"
  }, 
  "id": "582c4c93e20871aff6e898e7"
[\]
Running...

Task Failed

The importer docker_importer indicated a failed response when uploading
docker_tag unit to repository test2.

Deleting the upload request...
2016-11-16 13:09:56,558 - DEBUG - sending DELETE request to /pulp/api/v2/content/uploads/68ca05ad-468e-445d-bb83-fd1875058d0a/
2016-11-16 13:09:56,638 - INFO - DELETE request to /pulp/api/v2/content/uploads/68ca05ad-468e-445d-bb83-fd1875058d0a/ with parameters None
2016-11-16 13:09:56,638 - INFO - Response status : 200 

2016-11-16 13:09:56,639 - INFO - Response body :
 null

... completed

```

**Comment by daviddavis** [comment]
@ipanova good point. It looks like the exception is being caught here:

https://github.com/pulp/pulp_docker/blob/master/plugins/pulp_docker/plugins/importers/importer.py#L131

And then it's raised here:

https://github.com/pulp/pulp/blob/91a1e28c9e7d3dee418d5c7680dbf25c3e7adc63/server/pulp/server/managers/content/upload.py#L219-L224

This seems to work but I am not sure if it's the best solution?

https://gist.github.com/daviddavis/100c4e9463df7448d2b1a2c8a3161ac9

File: docs/user-guide/recipes.rst [hunk 2]
```diff
@@ -354,3 +354,63 @@ This produces a tarball at
...
+If we have a tag named latest and it points to the first manifest with digest
+sha256:4ecca..., we can point it to the second manifest with the following
+command::
+
+    $ pulp-admin docker repo tag --repo-id busybox --tag-name latest --manifest-digest sha256:c152ddeda2b828fbb610cb9e4cb121e1879dd5301d336f0a6c070b2844a0f56d
+
+We can also create a new tag and point it to the same manifest with::
+
+    $ pulp-admin docker repo tag --repo-id busybox --tag-name 1.2 --manifest-digest sha256:c152ddeda2b828fbb610cb9e4cb121e1879dd5301d336f0a6c070b2844a0f56d
+
```
Comment (ipanova) [inline docs/user-guide/recipes.rst]: maybe would be good to point out somewhere that the tag name references a single manifest?

File: docs/user-guide/release-notes/2.2.x.rst [hunk 3]
```diff
@@ -8,4 +8,9 @@ See the :fixedbugs:`2.2.0`.
 
 The docker rsync distributor has been improved to perform incremental publishes in cases where
 repositories have experienced only additive changes since the previous publish with the rsync
-distributor.
\ No newline at end of file
+distributor.
+
```
Comment (ipanova) [inline docs/user-guide/release-notes/2.2.x.rst]: if you write release notes for docker 2.2.x then this PR should be submitted against 2.2-dev branch and not master

File: docs/user-guide/release-notes/2.2.x.rst [hunk 4]
```diff
@@ -8,4 +8,9 @@ See the :fixedbugs:`2.2.0`.
...
+A new command, docker repo tag, can point tags to manifests. It accepts a
+`tag-name` and a `manifest-digest`. If a tag with the name `tag-name` does not
```
Comment (ipanova) [inline docs/user-guide/release-notes/2.2.x.rst]: s/tag-name/--tag-name/
s/manifest-digest/--manifest-digest/

Comment (ipanova) [inline docs/user-guide/release-notes/2.2.x.rst]: Also please mention `--repo-id` option, so user know the tag is created/updated within the repo

File: extensions_admin/pulp_docker/extensions/admin/pulp_cli.py [hunk 5]
```diff
@@ -11,7 +11,7 @@
 from pulp_docker.extensions.admin.images import ImageCopyCommand
 from pulp_docker.extensions.admin.images import ImageRemoveCommand
 from pulp_docker.extensions.admin.images import ImageSearchCommand
-from pulp_docker.extensions.admin.upload import UploadDockerImageCommand
+from pulp_docker.extensions.admin.upload import UploadDockerImageCommand, TagUpdateCommand
```
Comment (ipanova) [inline extensions_admin/pulp_docker/extensions/admin/pulp_cli.py]: please make imports in alphabetical order

File: plugins/pulp_docker/plugins/importers/importer.py [hunk 6]
```diff
@@ -126,8 +127,11 @@ def upload_unit(self, repo, type_id, unit_key, metadata, file_path, conduit, con
         :rtype:           dict
         """
         try:
-            upload_step = upload.UploadStep(repo=repo, file_path=file_path, config=config)
+            upload_step = upload.UploadStep(repo=repo, file_path=file_path, config=config,
+                                            metadata=metadata, type_id=type_id)
             upload_step.process_lifecycle()
+        except PulpCodedValidationException as e:
```
Comment (bmbouter) [inline plugins/pulp_docker/plugins/importers/importer.py]: You could just make this `except PulpCodedValidationException:` and then have L#134 be `raise`

**Comment by daviddavis** [comment]
ok test

**Comment by daviddavis** [comment]
@ipanova @dkliban I think I addressed all the comments if you want to re-review. Thanks.

File: docs/user-guide/recipes.rst [hunk 7]
```diff
@@ -354,3 +354,66 @@ This produces a tarball at
 both a JSON file for use with crane, and the static Image files to which crane
 will redirect requests. See the `Crane`_ documentation for how to use that
 tarball.
+
+Tagging a Manifest
+------------------
+
+Using the docker repo tag command, we can point a docker tag to a manifest. If
```
Comment (dkliban) [inline docs/user-guide/recipes.rst]: The command should be decorated with back ticks ``docker repo tag``.

File: docs/user-guide/release-notes/2.3.x.rst [hunk 8]
```diff
@@ -0,0 +1,12 @@
+2.3 Release Notes
+=================
+
+2.3.0
+-----
+
+See the :fixedbugs:`2.3.0`.
+
+A new command, docker repo tag, can point tags to manifests. It accepts
```
Comment (dkliban) [inline docs/user-guide/release-notes/2.3.x.rst]: The command should be decorated with double back ticks: ` ``docker repo tag`` `.

File: docs/user-guide/recipes.rst [hunk 9]
```diff
@@ -354,3 +354,66 @@ This produces a tarball at
...
+the tag we specify does not exist, it will be created. If the tag exists
+however, it will be updated as tag name is unique per repository and can point
+to only one manifest.
+
+For instance, suppose we have the following manifests::
+
+    $ pulp-admin docker repo search tag --repo-id busybox
```
Comment (ipanova) [inline docs/user-guide/recipes.rst]: i don't think this is what ` docker repo search tag` command displays, but `docker repo seach manifest`;
please update docs with correct info

File: docs/user-guide/release-notes/2.3.x.rst [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+`--repo-id`, `--tag-name`, and `--manifest-digest`. If a tag with the name
```
Comment (dkliban) [inline docs/user-guide/release-notes/2.3.x.rst]: The arguments should also be decorated with ``.

File: docs/user-guide/release-notes/2.3.x.rst [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+does not exist, it will be created. The tag is then updated and pointed to the
```
Comment (ipanova) [inline docs/user-guide/release-notes/2.3.x.rst]: maybe my English is not very good, but does not it miss:
'If the tag exist, it is then updated....'

Comment (dkliban) [inline docs/user-guide/release-notes/2.3.x.rst]: Good eye Ina.

s/The tag is then updated/If the tag exists, then it is updated/

**Comment by ipanova** [comment]
@daviddavis please address this comment https://github.com/pulp/pulp_docker/pull/184#pullrequestreview-9489284 and the one that was discussed on irc and good to merge from my side.

File: docs/user-guide/recipes.rst (hunk 2 above)
Comment (daviddavis) [inline docs/user-guide/recipes.rst]: I say something to that affect here:

https://github.com/pulp/pulp_docker/pull/184/files#diff-c8a41b4a5239ce5f909b7a6379e80040R363

**Comment by daviddavis** [comment]
@ipanova @dkliban thanks for the review. I updated the docs and added a note on the manifest tag field:

https://git.io/vXpgR

**Pull request merged**